node/src/da_block_import.rs

```diff
@@ -4,60 +4,94 @@
 /// to Babe and Grandpa.
 /// It double-checks the **extension header** which contains the `Kate Commitment` and `Data
 /// Root`.
-use std::sync::Arc;
-
 use avail_base::metrics::avail::ImportBlockMetrics;
 use avail_core::{
-	BlockLengthColumns, BlockLengthRows, HeaderVersion, OpaqueExtrinsic, BLOCK_CHUNK_SIZE,
+	ensure, header::HeaderExtension, BlockLengthColumns, BlockLengthRows, OpaqueExtrinsic,
+	BLOCK_CHUNK_SIZE,
 };
 use da_runtime::{
 	apis::{DataAvailApi, ExtensionBuilder},
 	Header as DaHeader,
 };
-use derive_more::Constructor;
-use frame_support::ensure;
 use frame_system::limits::BlockLength;
-use sc_client_api::Backend;
+
 use sc_consensus::{
 	block_import::{BlockCheckParams, BlockImport as BlockImportT, BlockImportParams},
 	ImportResult,
 };
-use sp_api::ProvideRuntimeApi;
+use sp_api::{ApiError, ProvideRuntimeApi};
 use sp_blockchain::HeaderBackend;
 use sp_consensus::{BlockOrigin, Error as ConsensusError};
 use sp_core::H256;
 use sp_runtime::traits::Block as BlockT;
+use std::{marker::PhantomData, sync::Arc, time::Instant};
 
-#[derive(Constructor)]
-pub struct BlockImport<BE, C, I> {
-	pub backend: Arc<BE>,
-	pub client: Arc<C>,
-	pub inner: I,
+pub struct BlockImport<B, C, I> {
+	client: Arc<C>,
+	inner: I,
 	// If true, it skips the DA block import check during sync only.
-	pub unsafe_da_sync: bool,
+	unsafe_da_sync: bool,
+	_block: PhantomData<B>,
 }
 
-impl<BE, C, I: Clone> Clone for BlockImport<BE, C, I> {
-	fn clone(&self) -> Self {
+impl<B, C, I> BlockImport<B, C, I>
+where
+	B: BlockT<Extrinsic = OpaqueExtrinsic, Header = DaHeader, Hash = H256>,
+	I: BlockImportT<B> + Clone + Send + Sync,
+	I::Error: Into<ConsensusError>,
+	C: ProvideRuntimeApi<B> + HeaderBackend<B> + Send + Sync,
+	C::Api: DataAvailApi<B> + ExtensionBuilder<B>,
+{
+	pub fn new(client: Arc<C>, inner: I, unsafe_da_sync: bool) -> Self {
 		Self {
-			backend: self.backend.clone(),
-			client: self.client.clone(),
-			inner: self.inner.clone(),
-			unsafe_da_sync: self.unsafe_da_sync,
+			client,
+			inner,
+			unsafe_da_sync,
+			_block: PhantomData,
 		}
 	}
+
+	fn ensure_valid_header_extension(
+		&self,
+		block: &BlockImportParams<B>,
+	) -> Result<(), ConsensusError> {
+		let block_len = extension_block_len(&block.header.extension);
+		let extrinsics = || block.body.clone().unwrap_or_default();
```

Comment:
Why does it need to be a closure?

Reply:
And yeah, I tried to make it a reference and just like you said things get indeed get problematic when we reach the `build_tx_data_from_opaque` function and try to  get the caller from ext.